packages/nuxt/src/core/nuxt.ts

```diff
@@ -197,8 +198,47 @@ async function initNuxt (nuxt: Nuxt) {
   }
 
   if (nuxt.options.dev) {
-    // Add plugin to check if layouts are defined without NuxtLayout being instantiated
-    addPlugin(resolve(nuxt.options.appDir, 'plugins/check-if-layout-used'))
+    // Add component usage detection
+    let hasBeenInitialized = false
+    const detectedComponents = new Set<string>()
+    const cache = new Set<string>()
+
+    const componentDetectionConfig = {
+      rootDir: nuxt.options.rootDir,
+      exclude: [
+      // Exclude top-level resolutions by plugins
+        join(nuxt.options.rootDir, 'index.html'),
+        // Keep only imports coming from the user's project (inside the rootDir)
+        new RegExp(`^(?!${escapeRE(nuxt.options.rootDir)}${escapeRE(sep)}).+[^\n]+$`)
+      ],
+      detectedComponents
+    }
+
+    addVitePlugin(() => DetectComponentUsagePlugin.vite(componentDetectionConfig))
+    addWebpackPlugin(() => DetectComponentUsagePlugin.webpack(componentDetectionConfig))
```

Comment:
We could probably use `addBuildPlugin`

Reply:
That's a good idea, thanks for the suggestion.

This pattern is repeated alot across the codebase (`addVitePlugin` + `addWebpackPlugin` instead of `addBuildPlugin`), maybe a future PR could refactor them to make plugin declaration consistent.